### PR TITLE
Add InMeleeRange

### PR DESCRIPTION
**Pull request by dev7355608**

### CODE REVIEW

File: LibThreatClassic2.lua [hunk 1]
```diff
@@ -696,6 +719,18 @@ function ThreatLib:OnEnable()
...
+	if class == "WARRIOR" or class == "DRUID" then
+		self:RegisterEvent("UPDATE_SHAPESHIFT_FORM")
+	elseif class == "PALADIN" or class == "PRIEST" or class == "SHAMAN" then
+		self:RegisterEvent("UNIT_AURA")
+	end
+
+	if class == "PRIEST" or class == "PALADIN" or class == "DRUID" or class == "SHAMAN" then
+		self:RegisterEvent("UNIT_SPELLCAST_SENT")
+	end
+
```
Comment (dfherr) [inline LibThreatClassic2.lua]: I don't really like this. This very much looks like something that should be handled within the class modules. (meaning the whole determination of current role)

File: LibThreatClassic2.lua [hunk 2]
```diff
@@ -852,29 +1023,46 @@ function ThreatLib:UpdatePartyGUIDs()
...
 
 function ThreatLib:UNIT_PET(event, unit)
-	if unit ~= "player" then return end
-	local exists = UnitExists("pet")
-	if exists and self.running then
-		self:DisableModule("Pet-r"..MINOR)
-		self:EnableModule("Pet-r"..MINOR)
-	else
-		self:DisableModule("Pet-r"..MINOR)
+	if unit == "player" then
```
Comment (dfherr) [inline LibThreatClassic2.lua:984]: why was this changed?

File: LibThreatClassic2.lua [hunk 3]
```diff
@@ -883,14 +1071,15 @@ local BLACKLIST_MOB_IDS = ThreatLib.BLACKLIST_MOB_IDS or {}
...
 	if not msg then return end
-	local guid, target_guid, val = strsplit(":", msg)
-	target_guid, val = strsplit("=", target_guid)
-	val = strsub(val, 1, -2)
-	if guid then
-		local dstGUID, threat = target_guid, tonumber(val)
+
+	local unitGUID, updates = strsplit(":", msg)
+
+	for _, update in ipairs({strsplit(",", updates)}) do
```
Comment (dfherr) [inline LibThreatClassic2.lua]: looks like this fixes a bug in multi-target threat updates. Can this be put into another PR?

File: LibThreatClassic2.lua [hunk 4]
```diff
@@ -1175,25 +1457,94 @@ do
...
+	elseif role == "HEALER" then
+		interval = 3.0
 	end
+
 	return interval
 end
 
+-- #NODOC
+function ThreatLib:GetPublishMeleeRangeInterval()
+	return 0.5
```
Comment (dfherr) [inline LibThreatClassic2.lua]: I'm not really a fan of wrapping constants in a function. Are these stubs? Otherwise use a "constant".

File: LibThreatClassic2.lua [hunk 5]
```diff
@@ -1107,9 +1198,59 @@ function ThreatLib:_clearThreat(guid)
...
+			default = true
+		end
+	end
+
+	if default == nil then
+		local _, class = UnitClass(unitID)
+
+		if class == "ROGUE" or class == "WARRIOR" then
+			default = true
+		elseif not UnitIsPlayer(unitID) and class ~= "MAGE" then
```
Comment (dfherr) [inline LibThreatClassic2.lua]: please add a comment that this catches all pets that are not mages (only known range pets have class mage)

File: LibThreatClassic2.lua [hunk 6]
```diff
@@ -711,6 +730,19 @@ end
...
 		guidLookup[UnitGUID("player")] = UnitName("player")
+
+		if not inRaid then
+			targetIDs["player"] = "target"
+
+			if UnitExists("pet") then
+				targetIDs["pet"] = "pettarget"
+			end
+
+			if self._setInMeleeRangeDefault then
```
Comment (dfherr) [inline LibThreatClassic2.lua]: what's the purpose of this if statement?

File: LibThreatClassic2.lua [hunk 7]
```diff
@@ -711,6 +730,19 @@ end
 function ThreatLib:PLAYER_ENTERING_WORLD(force)
 	if UnitGUID("player") then
 		guidLookup[UnitGUID("player")] = UnitName("player")
+
+		if not inRaid then
```
Comment (dfherr) [inline LibThreatClassic2.lua]: why are we only doing this, if the player is not in a raid?

File: LibThreatClassic2.lua [hunk 8]
```diff
@@ -852,29 +916,32 @@ function ThreatLib:UpdatePartyGUIDs()
...
 	self:GROUP_ROSTER_UPDATE()	--- Does gaining or losing a pet already fire this? Is this needed?
 end
 
 function ThreatLib:GROUP_ROSTER_UPDATE()
-	if timers.UpdateParty then
-		self:CancelTimer(timers.UpdateParty, true)
-		timers.UpdateParty = nil
-	end
-	timers.UpdateParty = self:ScheduleTimer("UpdateParty", 1.0)
+	self:UpdateParty()
```
Comment (dfherr) [inline LibThreatClassic2.lua]: please add the delay again. but with 0.5 sec instead of 1 (i falsely updated that once and forgot to turn it back)

File: LibThreatClassic2.lua [hunk 9]
```diff
@@ -1024,6 +1091,28 @@ function ThreatLib.OnCommReceive:SET_NPC_MODULE_VALUE(sender, distribution, var_
 	end
 end
 
+function ThreatLib.OnCommReceive:MELEE_RANGE_UPDATE(sender, distribution, msg)
+	if not msg then return end
+
+	local unitGUID = UnitGUID(sender)
+
+	for _, update in ipairs({strsplit(",", msg)}) do
+		local targetGUID, inMeleeRange = strsplit("=", update)
```
Comment (dfherr) [inline LibThreatClassic2.lua]: `string_gmatch(msg, "([^=:]+)=(%d+),")` is nicer imho

File: LibThreatClassic2.lua [hunk 10]
```diff
@@ -1284,21 +1500,48 @@ function ThreatLib:GetCumulativeThreat()
...
 ------------------------------------------------------------------------
-function ThreatLib:UnitInMeleeRange(unitID)
-	local meleeCheck
-	if GetItemInfo(8149) then
-		meleeCheck = IsItemInRange(8149, unitID) -- Voodoo Charm (5yd Range)
-	else
-		meleeCheck = CheckInteractDistance(unitID, 3)
+local IsItemInRange = _G.IsItemInRange
+function ThreatLib:UnitInMeleeRange(targetID)
+	return IsItemInRange(8149, targetID) -- Voodoo Charm (5yd Range)
```
Comment (dfherr) [inline LibThreatClassic2.lua]: this function needs hardening

the logic should be to fallback to CheckInteractDistance, if IsItemInRange returns false. 

e.g.

`return IsItemInRange(8149, targetID) or CheckInteractDistance(targetID, 3)`

File: LibThreatClassic2.lua [hunk 11]
```diff
@@ -1107,9 +1198,59 @@ function ThreatLib:_clearThreat(guid)
...
+
+	if meleeRange[guid] then
+		meleeRange[guid] = wipe(meleeRange[guid])
+	end
+
 	self:PublishThreat()
 end
 
+-- #NODOC
+function ThreatLib:_getInMeleeRange(unitGUID, targetGUID)
```
Comment (dfherr) [inline LibThreatClassic2.lua:1338]: I would prefer these methods to be PascalCase like most other methods. The few methods that also use this style need a refactor imho

File: LibThreatClassic2.lua (hunk 10 above)
Comment (dfherr) [reply LibThreatClassic2.lua]: This method could also use a meele range spell of warrior, rogue, paladin, druid and hunter for further hardeing. only priest, mage and warlock don't have meele range spells.

**Review by dfherr — CHANGES_REQUESTED** [review]

File: LibThreatClassic2.lua [hunk 12]
```diff
@@ -1578,6 +1821,82 @@ function ThreatLib:IsActive()
...
+
+		if not ThreatLib.meleeRangeCheckInterval then
+			ThreatLib.meleeRangeCheckInterval = ThreatLib:GetMeleeRangeCheckInterval()
+		end
+
+		if time - lastMeleeRangeCheckTime > ThreatLib.meleeRangeCheckInterval then
+			local threatTargets_player = ThreatLib.threatTargets[playerGUID]
+
+			if threatTargets_player then
+				local current, currentTarget = next(targetIDs)
```
Comment (dfherr) [inline LibThreatClassic2.lua]: maybe better curentUnitId and currentUnitIdTarget

File: LibThreatClassic2.lua [hunk 13]
```diff
@@ -1578,6 +1821,82 @@ function ThreatLib:IsActive()
...
+	local targets = {}
+
+	_eventFrame:SetScript("OnUpdate", function (self, elapsed)
+		local time = GetTime()
+
+		if not ThreatLib.meleeRangeCheckInterval then
+			ThreatLib.meleeRangeCheckInterval = ThreatLib:GetMeleeRangeCheckInterval()
+		end
+
+		if time - lastMeleeRangeCheckTime > ThreatLib.meleeRangeCheckInterval then
```
Comment (dfherr) [inline LibThreatClassic2.lua]: the whole block below should be moved to a method similar to the others imho. would make the code easier to understand

File: LibThreatClassic2.lua [hunk 14]
```diff
@@ -795,6 +822,27 @@ function ThreatLib:ThreatUpdatedForUnit(unit_id, target_guid, threat)
...
 		_callbacks:Fire("ThreatUpdated", unit_id, target_guid, threat)
+
+		if unit_id == playerGUID then
+			lastMeleeRangeCheckTime = 0
+		end
+	end
+end
+
+-- #NODOC
+function ThreatLib:MeleeRangeUpdated(unitGUID, targetGUID, inMeleeRange)
```
Comment (dfherr) [inline LibThreatClassic2.lua:871]: this method can be removed. It has the exact same interface as `MeeleRangeUpdatedForUnit`

File: LibThreatClassic2.lua [hunk 15]
```diff
@@ -1578,6 +1821,82 @@ function ThreatLib:IsActive()
...
+		end
+
+		if time - lastMeleeRangeCheckTime > ThreatLib.meleeRangeCheckInterval then
+			local threatTargets_player = ThreatLib.threatTargets[playerGUID]
+
+			if threatTargets_player then
+				local current, currentTarget = next(targetIDs)
+
+				for targetGUID in pairs(threatTargets_player) do
+					local target = targets[targetGUID]
```
Comment (dfherr) [inline LibThreatClassic2.lua]: maybe better targetUnidId

File: LibThreatClassic2.lua [hunk 16]
```diff
@@ -1578,6 +1821,82 @@ function ThreatLib:IsActive()
...
+		if not ThreatLib.meleeRangeCheckInterval then
+			ThreatLib.meleeRangeCheckInterval = ThreatLib:GetMeleeRangeCheckInterval()
+		end
+
+		if time - lastMeleeRangeCheckTime > ThreatLib.meleeRangeCheckInterval then
+			local threatTargets_player = ThreatLib.threatTargets[playerGUID]
+
+			if threatTargets_player then
+				local current, currentTarget = next(targetIDs)
+
```
Comment (dfherr) [inline LibThreatClassic2.lua:2040]:
```
-- for meele range checks we need to find targetUnitIds for all targetGUID a player has threat on
--
-- this is achieved by a loop over all targetGUIDs that tries to find party members 
-- that currently have a targetGUID in target. as the targets of a party member do not
-- change during a single loop invocation, other found targetUnitIds are cached, so 
-- we do not need to recheck every party member for subsequent targetGUIDs, but 
-- can simply check against the cache
````

File: LibThreatClassic2.lua [hunk 17]
```diff
@@ -1578,6 +1821,82 @@ function ThreatLib:IsActive()
...
+	_eventFrame:SetScript("OnUpdate", function (self, elapsed)
+		local time = GetTime()
+
+		if not ThreatLib.meleeRangeCheckInterval then
+			ThreatLib.meleeRangeCheckInterval = ThreatLib:GetMeleeRangeCheckInterval()
+		end
+
+		if time - lastMeleeRangeCheckTime > ThreatLib.meleeRangeCheckInterval then
+			local threatTargets_player = ThreatLib.threatTargets[playerGUID]
```
Comment (dfherr) [inline LibThreatClassic2.lua]: please don't mix variable styles `playerThreatTargets` would be in style

File: LibThreatClassic2.lua [hunk 18]
```diff
@@ -1578,6 +1821,82 @@ function ThreatLib:IsActive()
...
+						inMeleeRange = ThreatLib:UnitInMeleeRange(target)
+					end
+
+					ThreatLib:MeleeRangeUpdated(playerGUID, targetGUID, inMeleeRange)
+				end
+
+				wipe(targets)
+			end
+
+			local meleeRange_player = meleeRange[playerGUID]
```
Comment (dfherr) [inline LibThreatClassic2.lua]: `playerMeeleRanges`

File: LibThreatClassic2.lua [hunk 19]
```diff
@@ -1578,6 +1821,82 @@ function ThreatLib:IsActive()
...
+
+			if meleeRange_player then
```
Comment (dfherr) [inline LibThreatClassic2.lua]: what's the purpose of this loop? Setting the meelerange to nil on all targets that have no threat?

Why?

**Review by dfherr — CHANGES_REQUESTED** [review]

File: LibThreatClassic2.lua (hunk 19 above)
Comment (dev7355608) [reply LibThreatClassic2.lua]: This ensures that the range entries of targets that have died are deleted. We could do this in `_clearThreat` and `_clearAllThreat` alternatively.

File: LibThreatClassic2.lua [hunk 20]
```diff
@@ -795,6 +822,27 @@ function ThreatLib:ThreatUpdatedForUnit(unit_id, target_guid, threat)
...
+
+-- #NODOC
+function ThreatLib:MeleeRangeUpdatedForUnit(unitGUID, targetGUID, inMeleeRange)
+	if type(unitGUID) ~= "string" then
+		error(("Assertion failed: type(%s --[[unitGUID]]) == %q, trace: %s"):format(self:toliteral(unitGUID), "string", debugstack()))
+	end
+
+	if self:_getInMeleeRange(unitGUID, targetGUID) ~= inMeleeRange then
+		self:_setInMeleeRange(unitGUID, targetGUID, inMeleeRange)
+		_callbacks:Fire("MeleeRangeUpdated", unitGUID, targetGUID, inMeleeRange)
```
Comment (dfherr) [inline LibThreatClassic2.lua:883]: this callback is never used right?

File: LibThreatClassic2.lua (hunk 2 above)
Comment (dev7355608) [reply LibThreatClassic2.lua:984]: If the pet of someone other than the player changes, the group roster should be updated as well, no?

Comment (dfherr) [reply LibThreatClassic2.lua:984]: makes sense

File: LibThreatClassic2.lua (hunk 19 above)
Comment (dfherr) [reply LibThreatClassic2.lua]: How about adding it in another function and calling it here:
https://github.com/dfherr/LibThreatClassic2/blob/r6-alpha1/LibThreatClassic2.lua#L530-L533
that method could also use a rename (e.g. ScheduleCombatReset) or something.

File: LibThreatClassic2.lua (hunk 20 above)
Comment (dev7355608) [reply LibThreatClassic2.lua:883]: Melee range updates change the output of `UnitDetailedThreatSituation`. So anyone how want to be notified when this changes needs to register the `MeleeRangeUpdated` event.

File: LibThreatClassic2.lua (hunk 19 above)
Comment (dfherr) [reply LibThreatClassic2.lua]: Also shouldn't this actually remove the `targetGUID` from the players meele range table instead of setting it to nil.

File: LibThreatClassic2.lua (hunk 4 above)
Comment (dfherr) [reply LibThreatClassic2.lua]: Also this value should be much higher. At 800 CPS throttle and 39 characters per target (36 targetguid, 1 true/false, = and ,) plus the playerguid and message name, moving in and out of Garr adds range will instantly go in burst mode and cause throttle.

This should trigger every 5 seconds at most. Especially because it triggers instant on the first change.

File: LibThreatClassic2.lua (hunk 13 above)
Comment (dfherr) [reply LibThreatClassic2.lua]: it should also be moved to a scheduled timer rather than implementing that with time checks on every frame

File: LibThreatClassic2.lua [hunk 21]
```diff
@@ -1175,6 +1320,67 @@ do
 			self:SendCommRaw(self:GroupDistribution(), "THREAT_UPDATE", petMsg)
 		end
 		lastPublishTime = GetTime()
+
+		self:PublishMeleeRange(force)
```
Comment (dfherr) [inline LibThreatClassic2.lua]: please remove

**Review by dfherr — CHANGES_REQUESTED** [review]

File: LibThreatClassic2.lua [hunk 22]
```diff
@@ -795,6 +822,27 @@ function ThreatLib:ThreatUpdatedForUnit(unit_id, target_guid, threat)
 		self:UpdateTPS(unit_id, target_guid, threat)
 		self:_setThreat(unit_id, target_guid, threat)
 		_callbacks:Fire("ThreatUpdated", unit_id, target_guid, threat)
+
+		if unit_id == playerGUID then
+			lastMeleeRangeCheckTime = 0
+		end
```
Comment (dfherr) [inline LibThreatClassic2.lua:866]: please remove. this is unecessary.

**Review by dfherr — CHANGES_REQUESTED** [review]

File: LibThreatClassic2.lua (hunk 21 above)
Comment (dfherr) [reply LibThreatClassic2.lua]: meele range is only relevant for aggroMod calculations, which in turn are only relevant for threatPercentage calculations. It doesn't have to be included at so many places.

File: LibThreatClassic2.lua (hunk 7 above)
Comment (dev7355608) [reply LibThreatClassic2.lua]: Removed. It's redundant.

File: LibThreatClassic2.lua (hunk 10 above)
Comment (dev7355608) [reply LibThreatClassic2.lua]: Added `CheckInteractDistance`.

File: LibThreatClassic2.lua (hunk 19 above)
Comment (dev7355608) [reply LibThreatClassic2.lua]: Removed. It's in `clearThreat`/`_clearAllThreat` now. These functions should renamed in the future to reflect that: their purpose is really to clear any kind of data after leaving combat.

**Comment by Grookshank** [comment]
Is there any chance we could get this in the lib, so ranged and melee buffer would be different?

**Comment by dfherr** [comment]
this and tranquil totem are the next things on my list now that most pressing bugs are fixed.
